still_process fails spotfinding on image(s) in middle of range

Using https://zenodo.org/record/55058 as a test data set:
```
$ dials.stills_process test_1/test_1_0400.cbf.bz2
....
Error spotfinding test_1_0400.cbf 3
```
However, normal dials import works:
```
$ dials.import test_1/test_1_0400.cbf.bz2 && dials.find_spots imported.expt
...
Found 19 strong pixels on image 400
...
Saved 3 reflections to strong.refl
```

Debugging now.

```
dials.stills_process test_1/test_1_0400.cbf.bz2 squash_errors=False

Error spotfinding test_1_0400.cbf 3
Traceback (most recent call last):
  File "/net/dials/raid1/aaron/dialsbuild/build/../modules/dials/command_line/stills_process.py", line 1482, in <module>
    script.run()
  File "/net/dials/raid1/aaron/dialsbuild/build/../modules/dials/command_line/stills_process.py", line 647, in run
    do_work(0, iterable)
  File "/net/dials/raid1/aaron/dialsbuild/build/../modules/dials/command_line/stills_process.py", line 490, in do_work
    processor.process_experiments(item[0], item[1])
  File "/net/dials/raid1/aaron/dialsbuild/build/../modules/dials/command_line/stills_process.py", line 820, in process_experiments
    observed = self.find_spots(experiments)
  File "/net/dials/raid1/aaron/dialsbuild/build/../modules/dials/command_line/stills_process.py", line 902, in find_spots
    observed = flex.reflection_table.from_observations(experiments, self.params)
  File "/net/dials/raid1/aaron/dialsbuild/modules/dials/array_family/flex_ext.py", line 173, in from_observations
    return find_spots(experiments)
  File "/net/dials/raid1/aaron/dialsbuild/modules/dials/algorithms/spot_finding/finder.py", line 776, in __call__
    assert missed.count(True) == 0, missed.count(True)
AssertionError: Please report this error to dials-support@lists.sourceforge.net: 3
```

N.B. test fails with the change above:

```
Silver-Surfer command_line :) [master] $ pytest --regression test_find_spots.py::test_find_spots_from_imported_as_grid
Test session starts (platform: darwin, Python 3.6.7, pytest 4.6.4, pytest-sugar 0.9.2)
rootdir: /Users/graeme/svn/cctbx/modules/dials, inifile: pytest.ini
plugins: dials-data-2.0.69, sugar-0.9.2, mock-2.0.0, forked-1.1.2, xdist-1.31.0
collecting ... 

―――――――――――――――――――― test_find_spots_from_imported_as_grid ―――――――――――――――――――――

dials_data = <DataFetcher: /Users/graeme/svn/cctbx/build/dials_data>
tmpdir = local('/private/var/folders/hs/gvssfcd910s0jczjhjczj6h80000gn/T/pytest-of-graeme/pytest-12/test_find_spots_from_imported_0')

    def test_find_spots_from_imported_as_grid(dials_data, tmpdir):
        """First run import to generate an imported.expt and use this."""
        _ = procrunner.run(
            ["dials.import", "oscillation=0,0"]
            + [
                f.strpath for f in dials_data("centroid_test_data").listdir("centroid*.cbf")
            ],
            working_directory=tmpdir.strpath,
        )
    
        result = procrunner.run(
            [
                "dials.find_spots",
                tmpdir.join("imported.expt").strpath,
                "output.reflections=spotfinder.refl",
                "output.shoeboxes=True",
                "algorithm=dispersion",
            ],
            working_directory=tmpdir.strpath,
        )
        assert not result.returncode and not result.stderr
        assert tmpdir.join("spotfinder.refl").check(file=1)
    
        reflections = flex.reflection_table.from_file(tmpdir / "spotfinder.refl")
    
>       assert len(set(reflections["id"])) == 9, len(set(reflections["id"]))
E       AssertionError: 1
E       assert 1 == 9
E        +  where 1 = len({0})
E        +    where {0} = set(<scitbx_array_fami...ect at 0x1118ee870>)

test_find_spots.py:167: AssertionError
----------------------------- Captured stdout call -----------------------------
Warning: Ambiguous parameter definition: oscillation = 0,0
Best matches:
  input.tolerance.scan.oscillation
  geometry.scan.oscillation
Assuming geometry.scan.oscillation was intended.
DIALS (2018) Acta Cryst. D74, 85-97. https://doi.org/10.1107/S2059798317017235
DIALS 2.dev.1431-gb88160f3c
Warning: Ambiguous parameter definition: oscillation = 0,0
Best matches:
  input.tolerance.scan.oscillation
  geometry.scan.oscillation
Assuming geometry.scan.oscillation was intended.
The following parameters have been modified:

input {
  experiments = <image files>
}
geometry {
  scan {
    oscillation = 0 0
  }
}


Applying input geometry in the following order:
  1. Manual geometry

--------------------------------------------------------------------------------
  format: <class 'dxtbx.format.FormatCBFMiniPilatus.FormatCBFMiniPilatus'>
  num images: 9
  sequences:
    still:    1
    sweep:    0
  num stills: 0
--------------------------------------------------------------------------------
Writing experiments to imported.expt
DIALS (2018) Acta Cryst. D74, 85-97. https://doi.org/10.1107/S2059798317017235
DIALS 2.dev.1431-gb88160f3c
The following parameters have been modified:

output {
  reflections = "spotfinder.refl"
}
spotfinder {
  threshold {
    algorithm = *dispersion dispersion_extended
  }
}
input {
  experiments = /private/var/folders/hs/gvssfcd910s0jczjhjczj6h80000gn/T/pytest-of-graeme/pytest-12/test_find_spots_from_imported_0/imported.expt
}

Setting spotfinder.filter.min_spot_size=3
Configuring spot finder from input parameters
--------------------------------------------------------------------------------
Finding strong spots in imageset 0
--------------------------------------------------------------------------------


Finding spots in image 1 to 9...
Setting chunksize=9
Extracting strong pixels from images
 Using multiprocessing with 1 parallel job(s)

Found 1276 strong pixels on image 1
Found 1372 strong pixels on image 2
Found 1351 strong pixels on image 3
Found 1346 strong pixels on image 4
Found 1336 strong pixels on image 5
Found 1354 strong pixels on image 6
Found 1329 strong pixels on image 7
Found 1348 strong pixels on image 8
Found 1310 strong pixels on image 9

Extracted 3516 spots
Removed 1697 spots with size < 3 pixels
Removed 0 spots with size > 1000 pixels
Calculated 1819 spot centroids
Calculated 1819 spot intensities
Filtered 1812 of 1819 spots by peak-centroid distance

Histogram of per-image spot count for imageset 8:
1812 spots found on 8 images (max 301 / bin)
*      *
*      *
*      *
********
********
********
********
********
********
********
1      8

--------------------------------------------------------------------------------
Saved 1812 reflections to spotfinder.refl

 test/command_line/test_find_spots.py ⨯                          100% ██████████

Results (5.59s):
       1 failed
         - test/command_line/test_find_spots.py:142 test_find_spots_from_imported_as_grid
```

Trivial reproducer of error:

```
dials.stills_process squash_errors=0 convert_sequences_to_stills=1 ins_1d_grid_004?.cbf.gz
```

[exam.zip](https://github.com/dials/dials/files/4527272/exam.zip)

=> 

```
Grey-Area work :) $ dials.stills_process squash_errors=0 convert_sequences_to_stills=1 ins_1d_grid_004?.cbf.gz
Have 10 files
DIALS (2018) Acta Cryst. D74, 85-97. https://doi.org/10.1107/S2059798317017235
The following parameters have been modified:

dispatch {
  squash_errors = 0
}
geometry {
  convert_sequences_to_stills = 1
}

Loading files...
Loading ins_1d_grid_0040.cbf.gz
********************************************************************************
Finding Strong Spots
********************************************************************************
Setting spotfinder.filter.min_spot_size=3
Configuring spot finder from input parameters
--------------------------------------------------------------------------------
Finding strong spots in imageset 0
--------------------------------------------------------------------------------


Finding spots in image 1 to 1...
Setting chunksize=1
Extracting strong pixels from images
 Using multiprocessing with 1 parallel job(s)

Found 803 strong pixels on image 1

Extracted 46 spots
Removed 7 spots with size < 3 pixels
Removed 0 spots with size > 1000 pixels
Calculated 39 spot centroids
Calculated 39 spot intensities
Filtered 39 of 39 spots by peak-centroid distance
Error spotfinding ins_1d_grid_0040.cbf 39
Traceback (most recent call last):
  File "/Users/graeme/svn/cctbx/build/../modules/dials/command_line/stills_process.py", line 1482, in <module>
    script.run()
  File "/Users/graeme/svn/cctbx/build/../modules/dials/command_line/stills_process.py", line 647, in run
    do_work(0, iterable)
  File "/Users/graeme/svn/cctbx/build/../modules/dials/command_line/stills_process.py", line 563, in do_work
    processor.process_experiments(tag, experiments)
  File "/Users/graeme/svn/cctbx/build/../modules/dials/command_line/stills_process.py", line 820, in process_experiments
    observed = self.find_spots(experiments)
  File "/Users/graeme/svn/cctbx/build/../modules/dials/command_line/stills_process.py", line 902, in find_spots
    observed = flex.reflection_table.from_observations(experiments, self.params)
  File "/Users/graeme/svn/cctbx/modules/dials/array_family/flex_ext.py", line 173, in from_observations
    return find_spots(experiments)
  File "/Users/graeme/svn/cctbx/modules/dials/algorithms/spot_finding/finder.py", line 776, in __call__
    assert missed.count(True) == 0, missed.count(True)
AssertionError: Please report this error to dials-support@lists.sourceforge.net: 39
```

So we can at least look at this very easily.

--- a/algorithms/spot_finding/factory.py
+++ b/algorithms/spot_finding/factory.py
@@ -403,7 +403,7 @@ class SpotFinderFactory:
     """
 
     @staticmethod
-    def from_parameters(params=None, experiments=None):
+    def from_parameters(params=None, experiments=None, is_stills=False):
         """
         Given a set of parameters, construct the spot finder
 
@@ -464,6 +464,7 @@ class SpotFinderFactory:
             max_spot_size=params.spotfinder.filter.max_spot_size,
             no_shoeboxes_2d=no_shoeboxes_2d,
             min_chunksize=params.spotfinder.mp.min_chunksize,
+            is_stills=is_stills,
         )
 
     @staticmethod

--- a/algorithms/spot_finding/factory.py
+++ b/algorithms/spot_finding/factory.py
@@ -408,6 +408,8 @@ class SpotFinderFactory:
         Given a set of parameters, construct the spot finder
 
         :param params: The input parameters
+        :param is_stills:   [ADVANCED] Force still-handling of experiment
+                            ID remapping for dials.stills_process.
         :returns: The spot finder instance
         """
         if params is None:

--- a/algorithms/spot_finding/finder.py
+++ b/algorithms/spot_finding/finder.py
@@ -675,6 +675,7 @@ class SpotFinder:
         max_spot_size=20,
         no_shoeboxes_2d=False,
         min_chunksize=50,
+        is_stills=False,
     ):
         """
         Initialise the class.
@@ -703,6 +704,7 @@ class SpotFinder:
         self.mp_njobs = mp_njobs
         self.no_shoeboxes_2d = no_shoeboxes_2d
         self.min_chunksize = min_chunksize
+        self.is_stills = is_stills
 
     def __call__(self, experiments):
         warnings.warn(
@@ -744,7 +746,7 @@ class SpotFinder:
             for i, experiment in enumerate(experiments):
                 if experiment.imageset is not imageset:
                     continue
-                if experiment.scan:
+                if not self.is_stills and experiment.scan:
                     z0, z1 = experiment.scan.get_array_range()
                     z = table["xyzobs.px.value"].parts()[2]
                     table["id"].set_selected((z > z0) & (z < z1), i)

--- a/algorithms/spot_finding/finder.py
+++ b/algorithms/spot_finding/finder.py
@@ -683,6 +683,8 @@ class SpotFinder:
         :param find_spots: The spot finding algorithm
         :param filter_spots: The spot filtering algorithm
         :param scan_range: The scan range to find spots over
+        :param is_stills:   [ADVANCED] Force still-handling of experiment
+                            ID remapping for dials.stills_process.
         """
 
         # Set the filter and some other stuff

--- a/array_family/flex_ext.py
+++ b/array_family/flex_ext.py
@@ -136,7 +136,7 @@ class _(object):
         return result
 
     @staticmethod
-    def from_observations(experiments, params=None):
+    def from_observations(experiments, params=None, is_stills=False):
         """
         Construct a reflection table from observations.
 
@@ -167,7 +167,7 @@ class _(object):
         # Get the integrator from the input parameters
         logger.info("Configuring spot finder from input parameters")
         spotfinder = SpotFinderFactory.from_parameters(
-            experiments=experiments, params=params
+            experiments=experiments, params=params, is_stills=is_stills
         )
 
         # Find the spots

--- a/array_family/flex_ext.py
+++ b/array_family/flex_ext.py
@@ -142,6 +142,10 @@ class _(object):
 
         :param experiments: The experiments
         :param params: The input parameters
+        :param is_stills:   [ADVANCED] Force still-handling of experiment
+                            ID remapping for dials.stills_process. Do
+                            not use for general processing unless you
+                            know all the implications.
         :return: The reflection table of observations
         """
         from dials.algorithms.spot_finding.factory import SpotFinderFactory

--- a/command_line/stills_process.py
+++ b/command_line/stills_process.py
@@ -1021,7 +1021,9 @@ The detector is reporting a gain of %f but you have also supplied a gain of %f. 
         logger.info("*" * 80)
 
         # Find the strong spots
-        observed = flex.reflection_table.from_observations(experiments, self.params)
+        observed = flex.reflection_table.from_observations(
+            experiments, self.params, is_stills=True
+        )
 
         # Reset z coordinates for dials.image_viewer; see Issues #226 for details
         xyzobs = observed["xyzobs.px.value"]

--- a/newsfragments/1180.bugfix
+++ b/newsfragments/1180.bugfix
@@ -1,0 +1,1 @@
+``dials.stills_process``: Fix spotfinding error "Failed to remap experiment IDs"

--- a/test/command_line/test_stills_process.py
+++ b/test/command_line/test_stills_process.py
@@ -2,6 +2,7 @@ from __future__ import absolute_import, division, print_function
 
 import os
 
+import procrunner
 import pytest
 
 import dxtbx
@@ -162,17 +163,20 @@ def test_sacla_h5(dials_regression, run_in_tmpdir, use_mpi, in_memory=False):
     )
 
 
-def test_pseudo_scan(dials_data, run_in_tmpdir):
-    path = os.path.join(dials_data("centroid_test_data"), "centroid_000[1-2].cbf")
-
-    command = (
-        "dials.stills_process %s convert_sequences_to_stills=True squash_errors=False composite_output=True"
-        % path
+def test_pseudo_scan(dials_data, tmp_path):
+    result = procrunner.run(
+        (
+            "dials.stills_process",
+            dials_data("centroid_test_data") / "centroid_000[1-2].cbf",
+            "convert_sequences_to_stills=True",
+            "squash_errors=False",
+            "composite_output=True",
+        ),
+        working_directory=tmp_path,
     )
-    result = easy_run.fully_buffered(command).raise_if_errors()
-    result.show_stdout()
+    assert not result.returncode and not result.stderr
 
     experiments = ExperimentListFactory.from_json_file(
-        "idx-0000_refined.expt", check_format=False
+        tmp_path / "idx-0000_refined.expt", check_format=False
     )
     assert len(experiments) == 2

--- a/test/command_line/test_stills_process.py
+++ b/test/command_line/test_stills_process.py
@@ -160,3 +160,19 @@ def test_sacla_h5(dials_regression, run_in_tmpdir, use_mpi, in_memory=False):
             list(range(490, 515)),
         ],
     )
+
+
+def test_pseudo_scan(dials_data, run_in_tmpdir):
+    path = os.path.join(dials_data("centroid_test_data"), "centroid_000[1-2].cbf")
+
+    command = (
+        "dials.stills_process %s convert_sequences_to_stills=True squash_errors=False composite_output=True"
+        % path
+    )
+    result = easy_run.fully_buffered(command).raise_if_errors()
+    result.show_stdout()
+
+    experiments = ExperimentListFactory.from_json_file(
+        "idx-0000_refined.expt", check_format=False
+    )
+    assert len(experiments) == 2

